[IV-24-25] Objetivo 4

lmchaves opened the pull request:
* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
* [x]  ¿Se ha intentado cubrir con tests una parte de la lógica de negocio que se
   esté desarrollando?
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
* [x] ¿Estoy leyendo todo lo que marco o simplemente marco lo que se me pone por
delante?
* [x] ¿Se han documentado las elecciones de biblioteca de aserciones y del
   de *test runner*?
* [x] ¿Te has asegurado que lo que mencionas como biblioteca de aserciones y
test runners lo son realmente y tienen la misma funcionalidad? ¿O estás poniendo
`unittest` como *test runner* y `pytest` como biblioteca de aserciones?
* [x] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
* [x]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?
* [x] ¿Se han testeado también las excepciones?

lmchaves committed:
feat #24: agrego la elección del test runner

lmchaves committed:
agrego la clave para tests

lmchaves committed:
Update README.md

lmchaves committed:
feat(M1) #4: agrego funciones necesarias

Listaré las funciones: añadir_piezas, para poder añadir piezas al inventario

calcular_distancia: para saber las distancias entre dos sedes (para el costo del transporte)

crear_inventario: para que a partir de un excel con la pieza y el precio poder crear el inventario

lmchaves committed:
feat(M1) #4: agrego funciones necesarias a pedido.py

realizar_pedido: para poder hacer el pedido de piezas de una sede a otra

seleccionar_sede_optima: la implementación del algoritmo para encontrar la sede, que nos cueste menos la pieza (teneindo en ceunta el rpecio de estsa y el costo del tranporte)

lmchaves committed:
feat(M1) #4 #24: agrego los tests

test_realizar_pedido

test_seleccionar_sede_optima

lmchaves committed:
Update README.md

Al entregar me dio error: El README no contiene «make check test»

lmchaves committed:
Update README.md

Al entregar me dio error: El README no contiene «make check test»

lmchaves committed:
Eliminar archivo de master

lmchaves committed:
Merge branch 'Objetivo-4' of github.com:lmchaves/OrganizarTaller into Objetivo-4

lmchaves committed:
Merge branch 'master' into Objetivo-4

lmchaves committed:
pongo bien el aprtado de test

lmchaves committed:
Merge branch 'Objetivo-4' of github.com:lmchaves/OrganizarTaller into Objetivo-4

lmchaves committed:
me daba el sigueinte error al entregarlo:   Error: 🍋💥❌ El README no contiene «make check check»

hossam1522 commented on README.md:
```diff
@@ -17,11 +17,13 @@ La clase **Pedido** es responsable de gestionar las solicitudes de piezas para r
 ## Comprobación y tests
 Para ver si la sintaxis es correcta ejcuta:
-make check
+make check check
```
¿Y esto? ¿No debería ser simplemente `make check`?

lmchaves replied:
Es que al hacr la entrega me daba error al poenr solo make check, aqui lo corrijo: [4734153](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/473415373fc644f6edb3e57d54f71f2b3fc341c3)

hossam1522 replied:
Claro, es que en el iv.yaml solo deberías poner make y no make check, por eso te peta

hossam1522 commented on docs/test_runner.md:
```diff
@@ -0,0 +1,33 @@
+
+1. **Herramientas activas:** Preferencia por herramientas activamente mantenidas y actualizadas para evitar problemas de tecnologías obsoletas.
+
+2. **Pruebas automáticas:** ejecutar pruebas de forma automatizada.
```
Demasiado obvio como para ser un criterio, al final eso es lo que se busca en la búsqueda de una herramienta de este estilo

lmchaves replied:
ya esta corregido [4734153](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/473415373fc644f6edb3e57d54f71f2b3fc341c3)

hossam1522 commented on docs/test_runner.md:
```diff
@@ -0,0 +1,33 @@
+
+2. **Pruebas automáticas:** ejecutar pruebas de forma automatizada.
+
+3. **Biblioteca de aserciones**: biblioteca adecuada para verificar las condiciones esperadas en los tests.
```
Igual deberías expresarlo mejor. ¿Te refieres a que tenga aserciones la biblioteca? No se entiende demasiado bien realmente.

lmchaves replied:
aqui esta mejor explicado: [4734153](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/473415373fc644f6edb3e57d54f71f2b3fc341c3)

hossam1522 commented on docs/test_runner.md:
```diff
@@ -0,0 +1,33 @@
+
+
+
+Según los criterios establecidos, Pytest es la opción más adecuada debido a su facilidad de uso, mantenimiento activo, amplia comunidad y biblioteca de aserciones nativa, que simplifica la escritura y comprensión de los tests. Aunque Unittest es estándar, su curva de aprendizaje es más pronunciada, y Nose2, aunque versátil, tiene menor soporte. 
```
Has descartado Unittest porque su curva de aprendizaje es más pronunciada (aunque esto se podría llegar a considerar subjetivo). No deberías descartar una herramienta por algo que no está entre los criterios.

lmchaves replied:
ya lo he cambiado: [4734153](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/473415373fc644f6edb3e57d54f71f2b3fc341c3)

hossam1522 commented on src/pedido.py:
```diff
@@ -34,3 +36,56 @@ def __post_init__(self):
+    Returns:
+        Pedido: Objeto representando el pedido realizado.
+    """
+    # Calcular la distancia entre las sedes
```
Si el código no se entiende sin comentarios y te parece necesario añadirlos, estás incurriendo en deuda técnica. Aquí y en todo el resto de comentarios que tengas en el código.

lmchaves replied:
Ya lo he cmabiado para que el codigo se entienda sin los comnetarios : [2a133ca](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/2a133ca5d34fbb2446587ee00aaf36b32e29cbaf)

hossam1522 reviewed (commented):
Buenas! Se me ha asignado revisar tu issue. Te dejo algunos comentarios. ¡Ánimo!

MarioRgzLpz commented:
Buenas, se me ha asignado revisar tu PR. A parte de lo ya mencionado por @hossam1522 te dejo algunos comentarios.

MarioRgzLpz commented on docs/test_runner.md:
Aquí solo has tratado uno de los distintos niveles de abstracción que encontramos para las herramientas de testing. Debes añadir algo relacionado con las aserciones y las herramientas CLI. Además tienes algunas typos, intenta corregirlas.

lmchaves replied:
Aqui he añadirodo sobre las bibliotecas de aserciones usadsas en cda test runner y su uso en CLI: [b11186e](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/b11186ef34a3dee8ba1bb24dcfdb63b688664054)

MarioRgzLpz commented on docs/test_runner.md:
```diff
@@ -0,0 +1,33 @@
+
+## Pytest
+
+Permite ejecutar tests de foram automatizada mediante el comando **pytest**, es activamente mantenido y cuenta con una comunidad extensa, lo que asegura la resolución rápida de problemas (permitiendo tener muchos recursos).
```
```suggestion
Permite ejecutar tests de forma automatizada mediante el comando **pytest**, es activamente mantenido y cuenta con una comunidad extensa, lo que asegura la resolución rápida de problemas (permitiendo tener muchos recursos).
```
Typo

lmchaves replied:
ya eta corergido: [b11186e](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/b11186ef34a3dee8ba1bb24dcfdb63b688664054)

lmchaves committed:
fix(M1) #4 #24: arreglo los criterios

lmchaves committed:
corrgio la comprobacion

lmchaves committed:
fix(M1) #24: agergo información sobre las biblioteca de aserciones y CLI

lmchaves committed:
fix(M1) #4 #24: elimno los comntarios para no tener deuda tecnica

lmchaves commented:
Gracias @MarioRgzLpz y @hossam1522 por los comentarios, ya los he resuleto, si véis algo más comentadlo

lmchaves commented:
@JJ ya esta listo para revisión

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,49 @@
+Para seleccionar la herramietna para tests, se han considerado los siguientes criterios:
+
+
+1. **Herramientas activas:** Preferencia por herramientas activamente mantenidas y actualizadas para evitar problemas de tecnologías obsoletas.
```
Podrías considerar incluir la fuente de la que obtienes la información para validar este criterio.

lmchaves replied:
He añadido las fuentes [a5b4861](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/a5b4861d8a0328f2ec60527c0da4bf36a03f7524)

JJ replied:
Podías añadirla *aquí*

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,49 @@
+# Herramienta para tests
```
Ingual que ha comentado @MarioRgzLpz , deberías tratar los distintos niveles de abstracción. Buscando bibliotecas de aserciones, test runners y herramientas CLI, ¿solo has encontrado esas herramientas? Considera evaluar cada nivel de abstracción por separado y encontrarás más herramientas.

lmchaves replied:
Aqui he tratado mas a donof los distintos nievels, encontrando mas herramietas [a5b4861](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/a5b4861d8a0328f2ec60527c0da4bf36a03f7524)

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,49 @@
+
+1. **Herramientas activas:** Preferencia por herramientas activamente mantenidas y actualizadas para evitar problemas de tecnologías obsoletas.
+
+2. **Biblioteca de aserciones**: la herramienta debe incluir una biblioteca adecuada para verificar condiciones esperadas en los tests de manera clara y legible.
```
¿Qué es una biblioteca adecuada? Viendo la comparación que has hecho de las herramientas en cuanto a este criterio, hablas de que la herramienta ofrezca métodos que simplifiquen la aserción. Considera especificarlo en el criterio, ya que actualmente parece subjetivo.

lmchaves replied:
Lo especifico aqui: [a5b4861](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/a5b4861d8a0328f2ec60527c0da4bf36a03f7524)

juanmaaf reviewed (changes requested):
Buenas, se me ha asignado revisar tu PR. En general está bastante bien, pero he realizado una serie de puntualizaciones en la elección de las herramientas.

lmchaves committed:
fix(M1) #4 #24: incluyo fuentes, trato distintos niveles de abstraccion y modifico la descripción de criterios apra que sean objetivos

lmchaves commented:
Gracias @juanmaaf por los comentarios, ya los he resuelto

lmchaves commented:
@JJ ya está lsito apra revisión

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,81 @@
+
+## Criterios de evaluación
+
+Para seleccionar la herramienta para tests, se han considerado los siguientes criterios:
```
No existen las "herramientas para tests".

lmchaves replied:
Aqui esta modificado y con las fuentes en el criterio: [0ed63f1](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/0ed63f1b1633847d5be68e3d0ad6d5187314096b)

JJ reviewed (changes requested):
Por favor, lee los [errores comunes](https://github.com/JJ/IV-/blob/master/errores/objetivo-4.md#sobre-las-diferentes-herramientas):

> Hablar de "herramientas de test" sin especificar qué es lo que hacen esas herramientas de test. No existen las herramientas de test, existen herramientas que llevan a cabo diferentes funcionalidades relacionadas con los tests. El tipo de esas herramientas, para las funcionalidades que se requieren en el objetivo, es lo que hay que hay que reflejar aquí.

lmchaves committed:
fix(M1) #24: añado las fuentes en el criterio

lmchaves commented:
Ya esta listo @JJ

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,79 @@
+
+2. **Biblioteca de aserciones**: La herramienta debe incluir una biblioteca que permita verificar condiciones esperadas de los tests mediante métodos simples y legibles, proporcionando mensajes detallados en caso de fallo para facilitar la depuración.
+
+3. **Facilidad de ejecución desde CLI:** Capacidad para ejecutar pruebas fácilmente desde la línea de comandos, permitiendo ejecutar todos los tests o casos específicos.
```
Facilidad no es medible ni objetivo.

lmchaves replied:
Se ha eliminado el criterio al no ser objetivo [4d265d5](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/4d265d5c7094bbccb71cf21d1a5ae46964d17e3b)

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,79 @@
+
+## Conclusión
+
+Según los criterios establecidos, Pytest es la opción más adecuada. Aunque Unittest es una herramienta estándar y ampliamente usada, Pytest es preferido debido a su mantenimiento activo, amplia comunidad y capacidad para simplificar la escritura y comprensión de los tests. Nose2 y Testify, aunque versátiles, cuentan con menor soporte y comunidad en comparación.
```
En los criterios no hablas de comunidad, por lo que no deberías diferenciar herramientas por ello. La elección debe basarse en los criterios establecidos.

lmchaves replied:
Ya he modificado y en las conclusiones ya se usan los criterios [4d265d5](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/4d265d5c7094bbccb71cf21d1a5ae46964d17e3b)

juanmaaf reviewed (changes requested):
Dale una vuelta de nuevo a la elección de herramientas. Considera elegir varias herramientas si lo ves conveniente. No es necesario que la herramienta abarque los tres niveles de abstracción (test runner, biblioteca de aserciones y herramienta CLI para ejecutar los test).

lmchaves committed:
fix(M1) #24: modificación de cirterios

Se ha elminidao el criterio de CLi, yua que no era objetivo.

Se han añadido mas herramientas para biblioteca de aserciones.

Todas las conclusiones se basn en los criterios.

lmchaves commented:
@juanmaaf ya lo he resuelto

lmchaves commented:
@jj ya esta listo para revision

JJ commented:
> @JJ ya esta listo para revision

Espero a que lo revise @juanmaaf

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,88 @@
+
+## Conclusión
+
+Según los criterios establecidos, Pytest es la opción más adecuada debido a su mantenimiento activo, facilidad de uso y una biblioteca de aserciones integrada que simplifica la escritura y comprensión de los tests. Unittest sigue siendo una herramienta estándar confiable, aunque con una curva de aprendizaje más pronunciada. Nose2 y Testify ofrecen flexibilidad y extensibilidad, pero su menor nive ...
```
Hablar de curva de aprendizaje no es medible ni objetivo. Además, estás comparando herramientas en base a criterios no especificados en el apartado "Criterios de Evaluación" al principio del archivo.

lmchaves replied:
ya esta cambiado [37f7be9](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/37f7be9c00331acc8e1e241d1d2e022d50d1c723)

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,88 @@
+
+## Conclusión
+
+Pytest destaca por su simplicidad, mensajes automáticos detallados y mantenimiento activo. Ensure tiene una sintaxis explícita, mientras que Unittest permanece sólido para pruebas formales. PyHamcrest y Testify son útiles para expresiones avanzadas, pero menos intuitivos. Nose2 resulta menos práctico por su dependencia de bibliotecas externas.
```
Evita incluir términos no medibles ni objetivos como "simplicidad". ¿Qué herramienta es más simple y por qué?

lmchaves replied:
he eliminado los términos no ibjetivos ni medibles [37f7be9](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/37f7be9c00331acc8e1e241d1d2e022d50d1c723)

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,88 @@
+### [Testify](https://github.com/Yelp/Testify)
+
+- Los tests se ejecutan con el comando `testify` para todos los casos o especificando tests específicos mediante `testify ruta.TestClase.test_nombre`.
+
```
¿Qué comando eliges para ejecutar los tests?

lmchaves replied:
ya lo he incluido [37f7be9](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/37f7be9c00331acc8e1e241d1d2e022d50d1c723)

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,88 @@
+
+1. **Herramientas activas:** Preferencia por herramientas activamente mantenidas y actualizadas para evitar problemas de tecnologías obsoletas. Para ello se consulta en los repositorios oficiales de las herramientas, como [Pytest](https://github.com/pytest-dev/pytest), [Nose2](https://github.com/nose-devs/nose2), y [Testify](https://github.com/Yelp/Testify), donde se puede verificar la frecuencia ...
+
+2. **Biblioteca de aserciones**: La herramienta debe incluir una biblioteca que permita verificar condiciones esperadas de los tests mediante métodos simples y legibles, proporcionando mensajes detallados en caso de fallo para facilitar la depuración.
```
¿Consideras que este criterio se debe tener en cuenta para elegir el test runner y el comando para ejecutar los tests?

lmchaves replied:
ya he cmabiado los criterios para los fierentes niveles de abstracción [37f7be9](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/37f7be9c00331acc8e1e241d1d2e022d50d1c723)

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,88 @@
+1. **Herramientas activas:** Preferencia por herramientas activamente mantenidas y actualizadas para evitar problemas de tecnologías obsoletas. Para ello se consulta en los repositorios oficiales de las herramientas, como [Pytest](https://github.com/pytest-dev/pytest), [Nose2](https://github.com/nose-devs/nose2), y [Testify](https://github.com/Yelp/Testify), donde se puede verificar la frecuencia ...
+
+2. **Biblioteca de aserciones**: La herramienta debe incluir una biblioteca que permita verificar condiciones esperadas de los tests mediante métodos simples y legibles, proporcionando mensajes detallados en caso de fallo para facilitar la depuración.
+3. 
```
Elimina esta línea si no hay un tercer criterio.

lmchaves replied:
ya esta corregido [37f7be9](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/37f7be9c00331acc8e1e241d1d2e022d50d1c723)

juanmaaf reviewed (changes requested):
Buenas tardes. Disculpa la demora en la revisión. Te he dejado una serie de puntualizaciones, pues observo errores frecuentes en la elección de las herramientas. Considera separar los criterios de elección de herramientas para cada nivel de abstracción.

lmchaves committed:
fix(M1)#24: modifciación de cirterios y conlcuisones

se ha eliminado un cirterio no necesario, yañadido dos mas para los diferentes niveles de abstracción

lmchaves commented:
ya esta corregido @juanmaaf

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,94 @@
+
+1. **Herramientas activas:** Preferencia por herramientas activamente mantenidas y actualizadas para evitar problemas de tecnologías obsoletas. Para ello se consulta en los repositorios oficiales de las herramientas, como [Pytest](https://github.com/pytest-dev/pytest), [Nose2](https://github.com/nose-devs/nose2), y [Testify](https://github.com/Yelp/Testify), donde se puede verificar la frecuencia ...
+
+2. **Variedad de aserciones:** Se valoran bibliotecas que ofrecen una amplia gama de métodos de verificación, permitiendo cubrir diferentes casos de tests y simplificando la validación de condiciones complejas en las pruebas.
```
Esto es un error frecuente. Te invito a revisar mi PR y verás que he cometido el mismo error. No es medible.

lmchaves replied:
aqui esta corregido: [290c627](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/290c6275261a042c320285bb5f366d6aff6a0c08)

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,94 @@
+
+2. **Variedad de aserciones:** Se valoran bibliotecas que ofrecen una amplia gama de métodos de verificación, permitiendo cubrir diferentes casos de tests y simplificando la validación de condiciones complejas en las pruebas.
+   
+3. **Claridad de mensajes de error:** Es importante que proporcionen mensajes detallados y claros en caso de fallos, facilitando la identificación y solución de problemas durante el desarrollo de los tests.
```
Volvemos a lo mismo. "Claridad" no es medible ni obbjetivo.

lmchaves replied:
he modificado la parte de claridad , pero como la definicion del criterio si me parece correcto ya que habo de mensajes detallados que si son medibles y objetivos (claridad no, lo he modificado) [3540126](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/3540126ccd2192c05f565b688d8f010cf989552b)

juanmaaf commented on docs/test_runner.md:
```diff
@@ -0,0 +1,94 @@
+
+## Conclusión
+
+Según los criterios establecidos, Pytest es la opción más adecuada debido a su mantenimiento activo, facilidad de uso y una biblioteca de aserciones integrada que simplifica la escritura y comprensión de los tests. Unittest sigue siendo una herramienta estándar confiable así que tambien se puede usar en algunas ocasiones (se puede combianr con pytest). Nose2 y Testify ofrecen flexibilidad, pero s ...
```
La facilidad de uso no es un criterio. Además es algo completamente subjetivo. En la conclusión, donde eliges la herramienta, debes tener en cuenta únicamente los criterios que hayas establecido inicialmente. Deben ser criterios objetivos.

lmchaves replied:
ya he modificado la conlusión: [3540126](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/3540126ccd2192c05f565b688d8f010cf989552b)

juanmaaf reviewed (changes requested):
Has vuelto a cometer errores frecuentes. Revisa el archivo de errores frecuentes y te invito a revisar mi PR para que no cometas los errores que estoy cometiendo y estés más cerca de superar el objetivo.

lmchaves committed:
fix(M1) #24: modifico los cirterios

lmchaves committed:
fix(M1) #24: modificación de cirterio de bibilioteca

lmchaves commented:
@juanmaaf ya está lsito para revisión

juanmaaf reviewed (approved):
Creo que no tengo más que añadir. Por mi parte tienes el aprobado.

lmchaves commented:
@jj ya está lsito para revisión

JJ commented on README.md:
```diff
@@ -17,11 +17,13 @@ La clase **Pedido** es responsable de gestionar las solicitudes de piezas para r
+make check 
 
 Para poder realizar los test:
 make check test
```
¿Seguro que esto es así?

lmchaves replied:
Ya esta arreglado [1ee943d](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/1ee943d87f5e6bb00718cd3ad3c8fc8456e90949)

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,94 @@
+
+1. **Herramientas activas:** Preferencia por herramientas activamente mantenidas y actualizadas para evitar problemas de tecnologías obsoletas. Para ello se consulta en los repositorios oficiales de las herramientas, como [Pytest](https://github.com/pytest-dev/pytest), [Nose2](https://github.com/nose-devs/nose2), y [Testify](https://github.com/Yelp/Testify), donde se puede verificar la frecuencia ...
+
+2. Se valoran bibliotecas que ofrecen diversos métodos de verificación (por ejemplo, assertEqual, assertTrue, assertRaises, etc.), permitiendo cubrir diferentes casos de tests y simplificando la validación de condiciones complejas en las pruebas.
```
¿hay alguna que no los ofrezca? Por favor, lee [los errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/objetivo-3.md):

> Entre esos criterios arbitrarios, a veces hay criterios que se cumplen en cualquier caso y que no sirven para diferenciar uno de otros.

lmchaves replied:
ya he quitado este criterio, [2e6a5f4](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/2e6a5f461849caf8c275dc22e8e65b3aa84d3298)

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,94 @@
+
+2. Se valoran bibliotecas que ofrecen diversos métodos de verificación (por ejemplo, assertEqual, assertTrue, assertRaises, etc.), permitiendo cubrir diferentes casos de tests y simplificando la validación de condiciones complejas en las pruebas.
+   
+3. **Mensajes detallados de error:** Es importante que proporcionen mensajes detallados y claros en caso de fallos, facilitando la identificación y solución de problemas durante el desarrollo de los tests.
```
¿Qué mensajes de error y quién puede medir lo detallados que son?
Por otro lado, este "criterio" ya me lo he encontrado varias veces, por lo que entra en juego la primera regla del apartado "sobre los criterios"

lmchaves replied:
ya ha lo he arreglado [2e6a5f4](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/2e6a5f461849caf8c275dc22e8e65b3aa84d3298)

JJ reviewed (changes requested):
Agradezco a @juanmaaf el esfuerzo realizado, desgraciadamente, no supera el objetivo.

lmchaves committed:
fix(M1) #24: modifico los criterios de variedad y mensajes detallados

lmchaves committed:
modifico la errata para realizar los tests

lmchaves commented:
Ya he realiado als modifcaciones @JJ

JJ commented on README.md:
```diff
@@ -17,10 +17,12 @@ La clase **Pedido** es responsable de gestionar las solicitudes de piezas para r
 ## Comprobación y tests
 Para ver si la sintaxis es correcta ejcuta:
-make check
+make check 
```
Añadiéndole un espacio fuerzas un cambio, y por tanto la evaluación. Por favor, evitad esto.

JJ replied:
No se resuelve *nunca* sin ningún cambio. Y eso en el objetivo 4 deberíais saberlo ya.

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,92 @@
+
+1. **Herramientas activas:** Preferencia por herramientas activamente mantenidas y actualizadas para evitar problemas de tecnologías obsoletas. Para ello se consulta en los repositorios oficiales de las herramientas, como [Pytest](https://github.com/pytest-dev/pytest), [Nose2](https://github.com/nose-devs/nose2), y [Testify](https://github.com/Yelp/Testify), donde se puede verificar la frecuencia ...
+
+2. **Complejidad añadida:** Se penalizan herramientas que introduzcan complejidad innecesaria para el proyecto, ya sea por depender de bibliotecas adicionales o por requerir una sintaxis excesivamente compleja.
```
¿Con esto quieres decir que si viene incluida es mejor? Si es así, ¿porqué no pones eso en vez de poner un criterio que ni es medible ni, de hecho, comprensible?

lmchaves replied:
Aqui he cambiado el cirerio para que quede meor explicado [b4dd226](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/b4dd2265dfdb09ede50258ea371f7defd9089957)

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,92 @@
+
+### [Testify](https://github.com/Yelp/Testify)
+
+Testify es un test runner robusto, diseñado como una alternativa a Unittest. Proporciona una estructura clara para organizar tests y ofrece funcionalidades avanzadas como hooks para configurar y limpiar entornos. Aunque es menos popular que Pytest, es utilizado en proyectos que requieren flexibilidad adicional en sus pruebas.
```
¿Has visto lo que pone en la primera línea del README? ¿Qué tiene que ver con tu primer criterio?

lmchaves replied:
Ya lo he quitado al estar en desuso: [e80e6a3](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/e80e6a36eb3e94349bfd101038b753e74863edff)

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,92 @@
+
+## Conclusión
+
+Según los criterios establecidos, Pytest es la opción más adecuada debido a su mantenimiento activo, y una biblioteca de aserciones integrada que simplifica la escritura y comprensión de los tests. Unittest sigue siendo una herramienta estándar confiable así que tambien se puede usar en algunas ocasiones (se puede combinar con pytest). Nose2 y Testify  introducen un nivel de complejidad innecesar ...
```
¿Qué biblioteca de aserciones *integrada* tiene?
¿Qué pasa con la "complejidad"?

lmchaves replied:
Ya he modificado [56e23b3](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/56e23b357f0f9e5682747537650cb7c2e6f39315)

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,92 @@
+
+### [Pytest](https://github.com/pytest-dev/pytest)
+
+Pytest utiliza una biblioteca de aserciones nativa que permite escribir afirmaciones simples como `assert actual == expected`. Esta biblioteca genera automáticamente mensajes detallados en caso de fallos, lo que facilita la depuración y comprensión de errores.
```
No sólo no es `assert` parte de pytest, es que dijimos en clase que'es lo que era `assert`

lmchaves replied:
Ya ha sido arreglado [56e23b3](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/56e23b357f0f9e5682747537650cb7c2e6f39315)

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,92 @@
+
+### [Nose2](https://github.com/nose-devs/nose2)
+
+Nose2 no tiene su propia biblioteca de aserciones, pero es compatible con las bibliotecas estándar de Python como las de Unittest. Esto permite escribir pruebas utilizando métodos como `self.assertEqual` o incluso el operador `assert`.
```
O sea que estás incluyendo en las bibliotecas de aserciones una herramienta que no tiene su propia biblioteca de aserciones.

lmchaves replied:
Ya ha sido eliminado [56e23b3](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/56e23b357f0f9e5682747537650cb7c2e6f39315)

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,92 @@
+
+## Conclusión
+
+Pytest destaca por sus mensajes automáticos detallados y mantenimiento activo. Ensure tiene una sintaxis explícita, mientras que Unittest permanece sólido para pruebas formales y puede combinarse con Pytest para aprovechar lo mejor de ambos. PyHamcrest y Testify son útiles para expresiones avanzadas, pero menos intuitivos (pueden añadir complejidad innecesaria, ya que su sintaxis más elaborada).  ...
```
¿Qué se ha elegido?

lmchaves replied:
Lo he cambiado para que quede claro qeu se escoge [56e23b3](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/56e23b357f0f9e5682747537650cb7c2e6f39315)

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,92 @@
+
+### Unittest
+
+- Se utiliza el comando `python -m unittest` para ejecutar todos los tests, o `python -m unittest ruta.TestClase.test_nombre` para casos concretos.
```
Luego no es una herramienta CLI.

lmchaves replied:
Ya lo he modificado [56e23b3](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/56e23b357f0f9e5682747537650cb7c2e6f39315)

JJ commented on src/pedido.py:
```diff
@@ -34,3 +36,52 @@ def __post_init__(self):
+        piezas_requeridas=piezas_requeridas,
+        nombre_sede_origen=sede_origen.nombre,
+        nombre_sede_destino=sede_destino.nombre,
+        tiempo_estimado_llegada=distancia / 80,  
```
No se deben usar números mágicos. Ni aquí ni arriba.

lmchaves replied:
ya lo he resuelto: [6601616](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/66016166b8cd4c3ae31f3de193a2fc6bfe4ec709)

JJ commented on src/pedido.py:
```diff
@@ -34,3 +36,52 @@ def __post_init__(self):
+    Returns:
+        Pedido: Objeto representando el pedido realizado.
+    """
+    distancia = sede_origen.calcular_distancia(sede_destino)
```
Por el principio de responsabilidad única, realizar un pedido *solo* debe realizar un pedido, no calcular distancias.

lmchaves replied:
ahora realizar_pedido solo realiza dicho pedido [6601616](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/66016166b8cd4c3ae31f3de193a2fc6bfe4ec709)

JJ commented on src/pedido.py:
```diff
@@ -34,3 +36,52 @@ def __post_init__(self):
+            if pieza in sede.inventario and sede.inventario[pieza][0] >= cantidad_requerida:
+                distancia = sede_actual.calcular_distancia( sede)
+                costo_pieza = sede.inventario[pieza][1] * cantidad_requerida
+                costo_transporte = distancia * 0.5  
```
No solo estás haciendo esto con números mágicos, sino que estás repitiendo código aquí y arriba.

lmchaves replied:
ya esta arreglado [6601616](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/66016166b8cd4c3ae31f3de193a2fc6bfe4ec709)

JJ commented on src/pedido.py:
```diff
@@ -34,3 +36,52 @@ def __post_init__(self):
+                opciones.append((sede, costo_pieza + costo_transporte, distancia))
+        
+        if opciones:
+            sede_optima = min(opciones, key=lambda x: x[1])
```
¿Te parece que usar el primer elemento de un vector para almacenar el costo sea una forma expresiva de modelizar un problema? ¿No tendrá Python una forma mejor de hacerlo?

lmchaves replied:
ya lo he cambiado [6601616](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/66016166b8cd4c3ae31f3de193a2fc6bfe4ec709)

JJ commented on src/pedido.py:
```diff
@@ -34,3 +36,52 @@ def __post_init__(self):
+    for pieza, cantidad_requerida in piezas_requeridas.items():
+        opciones = []
+        for sede in sedes:
+            if pieza in sede.inventario and sede.inventario[pieza][0] >= cantidad_requerida:
```
Esto es un if dentro de un for que está dentro de un for. Para empezar, es complejo de testear por la complejidad ciclomática. Para seguir, hay herramientas mejores en Python para hacer este mismo código de forma más expresiva. Para continuar, no se comprende qué condición es lo que estás poniendo como primer elemento de un array. Si no se comprende, es imposible de depurar y de testear. Por favor, trata de escribir *tú mismo* un código que sea más comprensible, o revisar críticamente lo que te emita una herramienta generativa para escribir *código limpio* más fácil de entender por el revisor (y por ti)  y de testear.

lmchaves replied:
lo he cambiado para que sea mas limpio y facil de entender y no tanto for e if anidado: [6601616](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/66016166b8cd4c3ae31f3de193a2fc6bfe4ec709)

JJ reviewed (changes requested):
No he revisado los tests, pero por favor revísalos también mirando los errores frecuentes.

lmchaves committed:
fix(M1) #24: explico mejor el diterio de complejidad añadida

lmchaves committed:
fix(M1) #24: elimino testify ya que esta en desuso

lmchaves committed:
fix(M1) #24: realizo modficaciones tanto en biblioteca de arserciones como en herremienta cli

Primero modifco la conlcusión de test runner (con el cirterio de complejidad explicado correctametne

En bilbio teca de aserciones elimon aquellas qeu nol os son y modifoc la conclusion en consecuencia

En herramietnas cli elimno unnittest ya que utiliza python (el interprete)

lmchaves committed:
fix(M1) #4 #24: modifco la funcion realizar pedido para que no haya numeros magicos y que solo realice el pedido

lmchaves committed:
fix(M1) #4 #24: añado funciones neceasarias para pedido.py

lmchaves committed:
fix(M1) #4 #24: resuelve diferentes errores de realizar pedido y seleccionar la sede

elimnio los numeros mágicos

qeu realizr pedido solo realice el pedido

Que seleccionar sede optima sea menos complejo ciclomáticamente y tener un coódgo mas limipio siendo mas facil de entender

lmchaves committed:
fix(M1) #4 #24: modifco los test

lo modifico pra que cada test se centtre en testear una sola cosa y dar mensajes mas detallados

lmchaves commented:
@JJ ya he realizdo los cambios y he modificado errores que cometia en los tests

JJ commented on docs/test_runner.md:
```diff
@@ -4,7 +4,7 @@ Se han considerado los siguientes criterios, para las diferentes herramientas qu
 1. **Herramientas activas:** Preferencia por herramientas activamente mantenidas y actualizadas para evitar problemas de tecnologías obsoletas. Para ello se consulta en los repositorios oficiales de las herramientas, como [Pytest](https://github.com/pytest-dev/pytest), [Nose2](https://github.com/nose-devs/nose2), y [Testify](https://github.com/Yelp/Testify), donde se puede verificar la frecuencia ...
 
-2. **Complejidad añadida:** Se penalizan herramientas que introduzcan complejidad innecesaria para el proyecto, ya sea por depender de bibliotecas adicionales o por requerir una sintaxis excesivamente compleja.
+2. **Funcionalidades integradas:** Se valoran las herramientas que incluyan funcionalidades necesarias para las pruebas, evitando dependencias adicionales. Esto reduce la necesidad de instalar complementos externos, simplificando la configuración del entorno de pruebas.
```
De veras que debéis leer lo que escribís. ¿Hay alguna herramienta que *no* las incluya? ¿De veras es esto lo que quieres decir? ¿Que las herramientas de test incluyan funcionalidades necesarias para hacer tests?

JJ commented on docs/test_runner.md:
```diff
@@ -2,9 +2,9 @@
+1. **Herramientas activas:** Preferencia por herramientas activamente mantenidas y actualizadas para evitar problemas de tecnologías obsoletas. Para ello se consulta en los repositorios oficiales de las herramientas, como [Pytest](https://github.com/pytest-dev/pytest), [Nose2](https://github.com/nose-devs/nose2), donde se puede verificar la frecuencia de commits y actualizaciones recientes.
 
-2. **Funcionalidades integradas:** Se valoran las herramientas que incluyan funcionalidades necesarias para las pruebas, evitando dependencias adicionales. Esto reduce la necesidad de instalar complementos externos, simplificando la configuración del entorno de pruebas.
+2. **Funcionalidades integradas:** Se valoran las herramientas que incluyan una biblioteca de aserciones integrada y otras funcionalidades necesarias para las pruebas, evitando dependencias adicionales. Esto reduce la necesidad de instalar complementos externos, simplificando la configuración del entorno de pruebas.
```
Al menos lo has corregido, pero ¿estás diciendo que se valorará que las bibliotecas de aserciones incluyan una biblioteca de aserciones?

lmchaves replied:
Ya he dejado calro el cirterio [e946f51](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/e946f51b81ae7cc200a81b5f1d6bc6fb0dc9c330)

JJ commented on src/pedido.py:
```diff
@@ -37,27 +37,30 @@ def __post_init__(self):
                 raise ValueError(f"La cantidad de la pieza '{pieza}' debe ser mayor que cero.")
             
-def realizar_pedido(sede_origen, sede_destino, piezas_requeridas):
+def realizar_pedido(sede_origen, sede_destino, piezas_requeridas, distancia):
```
Por favor, sigue la regla de los 50 caracteres en la primera línea del mensaje de commit.

JJ reviewed (commented):
Por favor, mira los [errores comunes](https://github.com/JJ/IV-/blob/master/errores/objetivo-4.md):

> 

Los issues deben plantear problemas, igual que en el objetivo 2. El código resuelve el problema, y para ver si se ha resuelto, deben usarse los tests.

>    Un issue no puede ser "hacer tests"
>    No se puede plantear "escribir código que haga esto" como issue. Debe plantearse el problema que debe resolverse.

lmchaves committed:
closes #24 cambio el criterio de funcionalidades

lmchaves committed:
Update README.md el espacio

lmchaves commented:
@JJ como el #24 he mezclado tanto la selección de herramienta como el código, he creado el #26, pero tengo una duda, elimino los archivos de codigo y los vuelvo a poner para tener en #26 los menasjes de commit explicando el código? yaque njo lo he hecho bien

lmchaves committed:
closes #24: tengo que tener el codigo queresuelve el problema en #26

lmchaves committed:
closes #24: los test prueban si el probelma se ha resuelto correctamente

JJ commented:
> @JJ como el #24 he mezclado tanto la selección de herramienta como el código, he creado el #26, pero tengo una duda, elimino los archivos de codigo y los vuelvo a poner para tener en #26 los menasjes de commit explicando el código? yaque njo lo he hecho bien

haz lo que estimes conveniente y solicita evaluación cuando lo hagas. Yo no voy a decirte *nunca* lo que tienes que hacer.

lmchaves committed:
feat(M1) #26:resuelveo el problema

Primero, creo el metodo realizar_pedido, para poder simular el pedido entre sedes

seleccionar_sede_optima es la funcion que nos resuelve el problema de que sede escoger a al hora de realizar un oedido de piezas

lmchaves committed:
feat(M1) #26: agrego funciones necesarias

Añado funcones necesarias para la logica de negocio y para resolver el problema de selección de sedes

lmchaves committed:
feat(M1) #27: agergo los tests

Sea ralizan tests para comproabr qeu los pedidos e hagan de manera correcta

Tambien para comprobar si al sede elegidda es la mejor

lmchaves commented:
@JJ ya está listo pra revisión

JJ commented on src/pedido.py:
```diff
@@ -5,6 +5,8 @@
 
 from dataclasses import dataclass
 
+
```
Por favor, no metáis ruido con cambios sin sentido.

JJ replied:
Esto no está resuelto.

lmchaves replied:
perdón juraría que ya lo habia resuleto, aqui lo modifico [4a593ab](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/4a593ab12737990c71d45b86714cb6a69414a61b)

JJ commented on src/pedido.py:
```diff
@@ -30,7 +32,62 @@ def __post_init__(self):
+    Returns:
+        Pedido: Objeto representando el pedido realizado.
+    """
+    velcidad= 80  
```
typo. También hay una convención determinada para las constantes. Por favor, averíguala y úsala.

JJ reviewed (changes requested):
Tienes dos funciones, `realizar_pedido` y `seleccionar_sede_optima` que no tienen nada que ver la una con la otra. En la segunda tienes un pedidos.append pero no usa `realizar_pedido`. ¿Hay algo que me estoy perdiendo aquí? ¿Cuál de las dos es la lógica de negocio?

lmchaves committed:
fix #26:cambio realizar_pedido

Cambio realizar pedido para, ya qeu antes lo que yo hacia para la logica de negocia era primero llamar a la funcion seleccionar_sede_optima y despues a realizar_pedido, pero con el cambio ya no es necesario esta segunda llamada

lmchaves committed:
fix #26: defino correctmente la cte R

lmchaves committed:
fix #27: elimno los tests de realizar pedido

Ya que como hemos modifcado la funcion estos tests ya no teinene sentido

lmchaves committed:
fix #26: elimino ruido

lmchaves commented:
@JJ ya lo he cambiado aqui [0a46c6c](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/0a46c6c3ee0e7dfc1ad122f0edfe7fb13563fc2a)

lmchaves committed:
fix #27: agergo los datos para los tests

JJ commented on src/datos_sedes.py:
```diff
@@ -0,0 +1,50 @@
+from sede import Sede, crear_inventario
+
+# Cargar los datos de las sedes
+script_dir = os.path.dirname(__file__) # Obtiene el directorio del script actual
```
¿Te parece correcto que un fichero que se llama "datos" se cargue en el *código en producción*?

JJ reviewed (changes requested):
Esto no lo había visto antes, pero en infraestructura virtual lo que *nunca* se hace es incluir los datos en el código. Se trata precisamente de montar una infraestructura para que se carguen los datos cuando se necesiten, se arranque el programa o lo que sea.

JJ commented on src/datos_sedes.py:
```diff
@@ -0,0 +1,50 @@
+# datos_sedes.py
+import pandas as pd
```
No se puede usar ninguna biblioteca externa.

lmchaves committed:
fix #26: arreglo la llamada a realizar pedido

lmchaves committed:
fix #27: arreglo los test

ya que hemos cambiado tanto trealizar pedido como seleccionar la sede

lmchaves committed:
fix #27: elimno datos sedes

para lso datos de pureba lo realizo mediante un fixture

lmchaves committed:
Merge branch 'master' into Objetivo-4

lmchaves commented:
Ya lo he corregido y ya no cargo ningun dato @JJ : [fc1795c](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/fc1795cd5a2ff4c1b4625405a52dd0f580f012d4)

JJ reviewed (changes requested):
No se cierra *nunca* un comentario sin indicar cómo se ha resuelto (o resolviéndolo efectivamente haciendo un cambio)

lmchaves committed:
fix #26: elimino el ruido

lmchaves commented:
Ya lo he arreglado @JJ

JJ commented on src/pedido.py:
```diff
@@ -30,7 +33,62 @@ def __post_init__(self):
+                sede_origen=sede_actual,
+                sede_destino=mejor_opcion["sede"],
+                piezas_requeridas={pieza: cantidad},
+                distancia=mejor_opcion["distancia"],
```
¿Para qué quieres esto si ya has elegido la mejor opción?

JJ commented on src/pedido.py:
```diff
@@ -30,7 +33,62 @@ def __post_init__(self):
+        piezas_requeridas=piezas_requeridas,
+        nombre_sede_origen=sede_origen.nombre,
+        nombre_sede_destino=sede_destino.nombre,
+        tiempo_estimado_llegada=distancia / Pedido.VELOCIDAD,  
```
No recuerdo bien qué tienes en la HU requerida, pero no acabo de ver para qué necesitas esto en lo que es esencialmente un resultado de una función. ¿Haría falta siquiera esta función? Cuando la lógica de negocio produce un resultado, ¿no sería ya eso un producto mínimamente viable? Puede que esto lo necesites más adelante, pero esencialmente lo que tienes que ver es cuál es el PMV que necesitas.
Trabajar con PMVs simplifica mucho al evaluación por parte de los revisores, porque ven estricttamente lo que es necesario y pueden apreciar su calidad más fácilmente.

JJ commented on src/pedido.py:
Y realmente si esto es sólo necesario para almacenar el resultado de la lógica de negocio, tendrías que plantearte si *todo* es necesario.

lmchaves replied:
Tras pensarlo la verdad es qeu actualmente pedido solo almanceba el resultado, por lo que no hace mas que complicar el entendimiento, aqui lo he modificado [0a5e8aa](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/0a5e8aa571b3d9910d08ba7619ef8301a84f10a1)

JJ replied:
No entiendo muy bien. En ese commit borras el fichero, pero sigue apareciendo aquí.

JJ commented on src/sede.py:
```diff
@@ -16,18 +17,130 @@ class Sede:
     ubicacion_longitude: float
     inventario: dict[str, tuple[int, float]]
+    
+    R = 6371.0 
```
Wait, what? Una cosa es que no haya números mágicos y otra es que llames con un nombre mágico a un número mágico.

lmchaves replied:
ya lo he arreglado [0a5e8aa](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/0a5e8aa571b3d9910d08ba7619ef8301a84f10a1)

JJ commented on src/sede.py:
```diff
@@ -16,18 +17,130 @@ class Sede:
+            pieza (str): Nombre de la pieza para consultar (opcional).
+
+        Returns:
+            dict | tuple: Si `pieza` es None, devuelve el inventario completo como un diccionario.
```
Esto no puedes hacerlo NUNCA, usar valores específicos para significar una cosa diferente de lo que son. También incumple el principio de responsabilidad única.

JJ commented on tests/test_pedidos.py:
```diff
@@ -0,0 +1,97 @@
+@pytest.fixture
+def setup_data():
+    inventario_malaga = {
+        "Motor": (5, 3100),
```
Esto es problema del modelo, pero realmente el modelo debería de darle nombres específicos a valores específicos. Viendo esto quien revisa no puede saber si lo primero es el precio y lo segundo al cantidad, o al revés.

lmchaves replied:
aqui ya se entiede correctamente [021f75b](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/021f75b44588d5bd5ec2a358cfb3e526e4823664)

JJ reviewed (changes requested):
Aparte de las buenas prácticas, creo que hay algún problema en el modelo que deberías de reconsiderar. En caso necesario reabre los issues correspondientes, o crea issues nuevos.

lmchaves committed:
feat #34: elimno la función creada en #26

No es necesaria para el PMV

lmchaves committed:
feat #35: cambio la representacion creada en #10

En concreto a la hora de respresntear el inventario

lmchaves committed:
closes #35: elimino las funcones no ncesarias

lmchaves committed:
fix #26: elimino el ruido

lmchaves committed:
fix #27: modifico respecto a los cambios en el modelo

lmchaves committed:
Merge branch 'Objetivo-4' of github.com:lmchaves/OrganizarTaller into Objetivo-4

lmchaves committed:
feat #36: agrego la selección a sede

Elmino pedido.py yañado a sede la logica de negocio

lmchaves committed:
closes #27: ajusto los test a lso cambios del modelo

lmchaves commented:
ya está listo @JJ

JJ commented on src/sede.py:
```diff
@@ -15,19 +22,101 @@ class Sede:
             raise ValueError("El inventario debe ser un diccionario.")
 
-        # Validación de la ubicación: las coordenadas deben estar en [-90, 90] para la latitud y [-180, 180] para la longitud
         if self.ubicacion_latitude < -90 or self.ubicacion_latitude > 90:
```
Sigue habiendo números mágicos.

lmchaves replied:
Arreglado [8e6d21c](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/8e6d21cb29c38f3bfa9073da629e6250f536ae7a)

JJ commented on src/sede.py:
```diff
@@ -15,19 +22,101 @@ class Sede:
+
+        return distancia
+    
+    def tiene_pieza_disponible(self, pieza: str, cantidad_requerida: int) -> bool:
```
¿El nombre corresponde a lo que hace la función?

lmchaves replied:
Aqui ya esta corregido [505dc8c](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/505dc8c878b257861b701444a8c7e0ebbd7ee439)

JJ commented on src/sede.py:
```diff
@@ -15,19 +22,101 @@ class Sede:
+        Returns:
+            bool: True si la pieza está disponible en cantidad suficiente, False en caso contrario.
+        """
+        item = self.inventario.get(pieza)
```
¿Por qué usas `get` en vez de `[ ]`?

lmchaves replied:
Arreglado [8e6d21c](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/8e6d21cb29c38f3bfa9073da629e6250f536ae7a)

JJ commented on src/sede.py:
```diff
@@ -15,19 +22,101 @@ class Sede:
+
+        Args:
+            sede_destino (Sede): La sede de destino a la que se enviarán las piezas.
+            coste_km (float): Costo por kilómetro de transporte. Por defecto es 0.5.
```
El problema de los comentarios, aunque sean docstrings, es que hay que tratarlos como el código y mantenerlos de la misma manera. Además, si el comentario no se corresponde con el código, *los dos están mal*

JJ commented on src/sede.py:
```diff
@@ -15,19 +22,101 @@ class Sede:
+        Returns:
+            float: El costo de transporte calculado.
+        """
+        coste_km=0.5
```
¿Esto está declarado correctamente según las buenas prácticas de Python? ¿Incluyendo el lugar?
Al final, esta función hace una operación. ¿De veras merece la pena tener una función para esto?

lmchaves replied:
Ya lo he declarado correctamente y he decidido no mantener la función [8b3e4a6](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/8b3e4a63fee06c6727b1d8139d2a0b14cb9cdd7e)

JJ commented on src/sede.py:
```diff
@@ -15,19 +22,101 @@ class Sede:
+                )
+            )
+
+            if opciones:
```
Un if después de un map y un filter... ¿No se podría simplificar? Todo ello dentro de un `for`, además...

lmchaves replied:
Lo he simplificado [505dc8c](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/505dc8c878b257861b701444a8c7e0ebbd7ee439)

JJ commented on src/sede.py:
```diff
@@ -15,19 +22,101 @@ class Sede:
+
+            if opciones:
+                mejor_opcion = min(opciones, key=lambda x: x["costo_total"])
+                pedidos.append({
```
¿Te parece que según el principio de la responsabilidad única una función que se llama `seleccionar_sede_óptima` añada a la lista de pedidos?

lmchaves replied:
Lo he arreglado [505dc8c](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/505dc8c878b257861b701444a8c7e0ebbd7ee439)

JJ reviewed (changes requested):
Todavía hay algunos aspectos que podrías mejorar. Ánimo.

lmchaves committed:
fix #26: arreglo nombre de funciones y variables

Elminio los números mágicos, pongo bien el nombre a las funcioens para que representan su uso correctamente

lmchaves committed:
closes #26:respeto principio de la responsabilidad única

EN lugar de que seleccionar_sede_optima seleccione la sede y añada a la lsita de pedidos, creo otra función generar pedido la cual se encargue de de la lista de pedidos, dejanod en en seleccionar_sede_optima, que seleeccione aquella sede que minimice el costo, simplificando el for, map.filter e if que tenia antes

lmchaves committed:
closes #26: elmino la funcin ya que solo es una operación

lmchaves commented:
Ya he realizado las correcciones @JJ

JJ commented on src/sede.py:
```diff
@@ -15,19 +22,91 @@ class Sede:
+        if self.ubicacion_latitude < -Sede.LATIUD or self.ubicacion_latitude > Sede.LATIUD:
             raise ValueError("La latitud debe estar en el rango [-90, 90].")
-        if self.ubicacion_longitude < -180 or self.ubicacion_longitude > 180:
+        if self.ubicacion_longitude < -Sede.LONGITUD or self.ubicacion_longitude > Sede.LONGITUD:
```
De veras, Luis, desde el cariño, ¿crees que la longitud y latitud máxima son variables *de sede*?

lmchaves replied:
ya esta solucionado [6d9a167](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/6d9a167ea60d8ff4f9a50b468de85088c848fea6)

JJ commented on src/sede.py:
```diff
@@ -15,19 +22,91 @@ class Sede:
+
+        dlat = radians(otra_sede.ubicacion_latitude - self.ubicacion_latitude)
+        dlon = radians(otra_sede.ubicacion_longitude - self.ubicacion_longitude)
+        a = sin(dlat / 2) ** 2 + cos(radians(self.ubicacion_latitude)) * cos(radians(otra_sede.ubicacion_latitude)) * sin(dlon / 2) ** 2
```
Las IA generativas tienen la mala costumbre de usar variables sin nombres significativos. No tiene mayor importancia porque esta no es la parte principal de la lógica de negocio, pero de verdad que deberíais usar identificadores con nombres significativos.

lmchaves replied:
ya esta arreglado [6d9a167](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/6d9a167ea60d8ff4f9a50b468de85088c848fea6)

JJ commented on src/sede.py:
```diff
@@ -15,19 +22,91 @@ class Sede:
+    def generar_pedidos(piezas_requeridas, sedes, sede_actual):
+        pedidos = []
+        for pieza, cantidad in piezas_requeridas.items():
+            mejor_opcion = Sede.seleccionar_sede_optima(pieza, cantidad, sedes, sede_actual)
```
Luis, te prometo que tenía la mejor voluntad de dejarte pasar en esta ronda. Pero si tomas una función que no tiene una sola responsabilidad y la divides en dos funciones **una de las cuales hace dos cosas, generar pedidos y otra cosa que no dice  en el nombre, que es seleccionar sede óptima**, es que no has entendido qué es el principio de responsabilidad única. Por favor, trata de entenderlo  y crea un flujo de datos dentro del programa con una serie de funciones que trabajen con los datos *solo de una forma* y que *no tengan efectos secundarios* que los hacen difíciles o imposibles de depurar.

lmchaves replied:
Aqui esta solucionado y ya no forma parte de sede: [62fa512](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/62fa512ec35f4e266e025c6ccaf2fce6e5f90aad)

JJ commented on src/sede.py:
```diff
@@ -15,19 +22,91 @@ class Sede:
+        return min(opciones, key=lambda x: x["costo_total"], default=None)
+    
+    @staticmethod
+    def generar_pedidos(piezas_requeridas, sedes, sede_actual):
```
Por otro lado, estás trabajando con funciones estáticas, varias de las cuales reciben como argumento `sede`. Un objeto de una clase es simplemente una colección de funciones en las que uno de los argumentos es siempre el mismo. ¿Puede ser que teniendo varias funciones estáticas (es decir, que no son parte de un objeto) y que además tengan todas un argumento común, `sedes`, lo que esté pidiendo es que trates de modelar esto de una forma un poco más adecuada? ¿Que te permita además incluir un flujo de trabajo con funciones cada una de las cuales tenga una responsabilidad única?

lmchaves replied:
aqui lo he elminiado de sede [50d2bda](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/50d2bdab8699ea67c8d5ee58366b8b4aee7b2e60) y lo he cambiado para que los gestione gestorPedidos [62fa512](https://github.com/lmchaves/OrganizarTaller/pull/25/commits/62fa512ec35f4e266e025c6ccaf2fce6e5f90aad)

JJ reviewed (changes requested):
En general, Luis, trabajas correctamente y sueles tener una actitud positiva con los problemas que te señalan. El problema es que eres reactivo y no ves si los problemas resueltos no causan algún problema adicional, o simplemente has solucionado el problema en un lado e introducido en otro.
Ese es el principal problema, y si eso no hubiera estado ahí, no te habría indicado nada de la modelización del problema. Pero quizás refactorizando para modelizar mejor el dominio del problema puedas llegar a una mejor solución para los flujos de datos que incluyan, esta vez, el seguimiento del principio de responsabilidad única.

lmchaves committed:
fix #26_ modifoc nombres y constantes

lmchaves committed:
fix #26 #34: eliminamos las funciones estaticas

Ya que estas funciones teinen como praremtros sedes nos indican que nuestro modelo, no es correcto por eso, se crea pedido y gestorPedidos

lmchaves committed:
fix #26 #34: se añade pedido

lmchaves committed:
closes #26 #34: se añade getorPedidos

El cual realiza la logica de negocio, sienod el centro de la aplicación

lmchaves committed:
closes #26 #34 elmino ruido

lmchaves committed:
closes #34: soluciono un conflicto

lmchaves commented:
Lo primero @JJ gracias por la paciencia, y he decidido modificar el modelo añadaiendo la entidad gestorPedidos la cual se encarga de la lógica de negocio, para tener así un correcto flujo de datos y cumplir el principio de responsabilidad unica.

JJ commented on tests/test_pedidos.py:
```diff
@@ -0,0 +1,98 @@
+
+    assert len(resultado) == 3, f"Se esperaba 3 piezas, pero obtuve {len(resultado)}"
+
+    pedido_motor = next((p for p in resultado if p["pieza"] == "Motor"), None)
```
Aquí podrías aplicar un poco de DRY.

JJ reviewed (approved):
Aunque todavía podría mejorar, yo creo que el objetivo está superado.
Por favor, recuerda que las revisiones de compañeros son importantes para los 2 puntos adicionales; aunque no hayas salido aleatoriamente y aunque haya revisado yo ya, mira sobre todo en el objetivo 4 aquellos que estén activos (que hayan solicitado revisión en las últimas dos semanas) y revisa alguno. Puedes verlos en la página donde se entregan los objetivos.

lmchaves committed:
closes #27: mejoro los tests

lmchaves committed:
Merge branch 'master' into Objetivo-4

lmchaves committed:
closes #27
